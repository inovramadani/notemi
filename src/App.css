* {
	box-sizing: border-box;
}

.App {
	margin: 5vh;
	position: relative;
}

.center {
  text-align: center;
}

.input-note {
	display: inline-block;
}

.notes {
	text-align: center;
}

#btn-save {
	position: absolute;
	right: 20px;
}

#input-new-note {
	padding: 8px;
	margin-bottom: 1px;
}