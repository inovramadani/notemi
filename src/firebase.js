// Initialize Firebase
import firebase from 'firebase'

const config = {
  apiKey: process.env.REACT_APP_apiKey,
  authDomain: process.env.REACT_APP_authDomain,
  databaseURL: process.env.REACT_APP_databaseURL,
  projectId: process.env.REACT_APP_projectId,
  storageBucket: process.env.REACT_APP_storageBucket,
  messagingSenderId: process.env.REACT_APP_messagingSenderId
}

const settings = {timestampsInSnapshots: true}

firebase.initializeApp(config)
firebase.firestore().settings(settings)

export const db = firebase.firestore().collection('notes')
export const docName = 'user1'
export default firebase