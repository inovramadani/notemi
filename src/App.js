import React, { Component } from 'react'
import queryString from 'query-string'

import { db } from './firebase'

import './App.css'
import Button from './components/Button'
import Note from './components/Note'
import ColorPicker from './components/ColorPicker'

let docName = 'user1'
class App extends Component {
  state = {
    title: 'Notemi',
    text: '',
    colors: ['blue', 'grey', 'green'],
    selectedColor: 'blue',
    notes: []
  }

  render() {
    const { 
      title, 
      text,
      selectedColor,
      colors,
      notes 
    } = this.state

    return (
      <div className="App">
        <h1 className="title center">{title}</h1>
        <Button id="btn-save" onClick={this.handleClickSave}>Save</Button>
        <div className="input-note">
          <input id="input-new-note" value={text} onChange={this.handleChangeText} />
          <Button id="btn-add" onClick={this.handleClickAddNote}>Add note</Button>
        </div>
        <ColorPicker colors={colors} selectedColor={selectedColor} onChange={this.handleChangeColor} />
        <div className='notes'>
        { 
          notes.map((note, index) => (
            <Note 
              key={index}
              color={note.color} 
              text={note.text} 
              onChange={event => this.handleChangeNoteText(event, index)}
              onDelete={event => this.handleClickDeleteNote(index)}
            />
          )) 
        }
        </div>
      </div>
    )
  }

  componentWillMount() {
    const { user } = queryString.parse(window.location.search)
    if (user) docName = user
    this.loadData()
  }

  loadData() {
    return new Promise((resolve, reject) => {
      db.get().then((querySnapshot) => {
        let notes = []
        querySnapshot.forEach((doc) => {
          if (doc.id === docName) {
            notes = doc.data().notes
          }
        })

        this.setState({ notes })
        resolve('success')
      }).catch(error => {
        console.error(error)
        reject('failed')
      })
    })
  }

  handleClickSave = () => {
    this.saveData()
  }

  saveData() {
    return new Promise((resolve, reject) => {
      const { notes } = this.state     
      
      db.doc(docName).set({notes})
      .then(() => {
        window.alert('notes saved in database')
        resolve('success')
      })
      .catch(error => {
        console.error(error)
        reject('failed')
      })      
    })
  }

  handleChangeText = ({ target }) => {
    this.setState({ text: target.value })
  }

  handleChangeColor = ({ target }) => {
    this.setState({ selectedColor: target.value })
  }

  handleClickAddNote = () => {
    const { selectedColor: color, text, notes } = this.state
    const newNote = { color, text }
    const newNotes = [...notes, newNote]

    this.setState({ notes: newNotes })
  }

  handleChangeNoteText = ({ target }, index) => {
    const { notes } = this.state

    notes[index].text = target.value

    this.setState({ notes })
  }

  handleClickDeleteNote = (index) => {
    const { notes } = this.state
    const newNotes = [...notes]

    newNotes.splice(index, 1)
    this.setState({ notes: newNotes })
  }
}

export default App;
