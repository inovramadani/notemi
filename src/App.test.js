import React from 'react'
import ReactDOM from 'react-dom'
import App from './App'
import {shallow, mount, render} from 'enzyme'
import renderer from 'react-test-renderer'
import firebase from './firebase'
import db, { data_db_mock } from './db_mock'

describe('<App />', () => {
	it('renders without crashing', () => {
	  const div = document.createElement('div')
	  ReactDOM.render(<App />, div)
	  ReactDOM.unmountComponentAtNode(div)
	})

	it('renders correctly', () => {
		const app = renderer.create(<App />).toJSON()
		expect(app).toMatchSnapshot()
	})

	it('should be able to add note', () => {
		const app = shallow(<App />)
		app.find('#btn-add').simulate('click')
		app.find('#btn-add').simulate('click')
		app.find('#btn-add').simulate('click')
		app.find('#btn-add').simulate('click')
		expect(app.state().notes.length).toBe(4) // check on state

		app.update()
		expect(app.find('.notes').children()).toHaveLength(4) // check on node
	})

	it('should be able to edit note', () => {
		const app = mount(<App />)
		app.setState({notes: [{color: 'blue', text: 'hello'}]})
		app.find('.note-text').simulate('change', {target: {value: 'hallo'}})
		expect(app.state().notes[0].text).toBe('hallo') // check on state

		app.update()
		expect(app.find('.note-text').text()).toBe('hallo') // check on node
		app.unmount()
	})

	it('should be able to delete note on state (1)', () => {
		const app = shallow(<App />)
		app.find('#btn-add').simulate('click') // add one note
		app.instance().handleClickDeleteNote(0) // delete the note
		expect(app.state().notes).toEqual([])
	})

	it('should be able to delete note on state (2)', () => {
		const app = mount(<App />)
		app.setState({notes: [{color: 'blue', text: 'hello'}]}) // add one note
		app.find('.icon-delete').simulate('click') // delete the note
		expect(app.state().notes).toEqual([])
		app.unmount()
	})

	it('should be able to delete note on node', () => {
		const app = shallow(<App />)
		app.instance().handleClickAddNote() // add one note
		app.instance().handleClickDeleteNote(0) // delete the note
		expect(app.find('.notes').children()).toHaveLength(0)
	})

	it('should delete correct item in the notes array', () => {
		const app = mount(<App />)

		// check on state
		const notes = [{color: 'blue', text: '1'}, {color: 'green', text: '2'}, {color: 'grey', text: '3'}]
		app.setState({notes})
		app.instance().handleClickDeleteNote(1) // delete note {color: 'green', text: '2'}

		const updatedNotes = [{color: 'blue', text: '1'}, {color: 'grey', text: '3'}]
		expect(app.state().notes).toEqual(updatedNotes) 

		app.update()

		// check on node
		const node_notes = app.find('.notes')
		expect(node_notes.children()).toHaveLength(2)

		const node_colors = app.find('.note').getElements().map(el => el.props.style)
		expect(node_colors).toEqual([{backgroundColor: 'lightblue'}, {backgroundColor: 'lightgrey'}])

		const node_texts = app.find('.note-text').getElements().map(el => el.props.value)
		expect(node_texts).toEqual(['1','3'])
		
		app.unmount()
	})

	it('should be success to load notes from database', () => {
		const app = shallow(<App />)
		return app.instance().loadData().then(res => {
			expect(res).toBe('success')
		})
	})

	it('should be able to load mock data (mock firebase API call)', () => {
		jest.mock('./firebase')
		// const getDataSpy = jest.spyOn(firebase.firestore().collection('notes'), 'get')
		firebase.firestore().collection('notes').get = jest.fn().mockImplementation(() => data_db_mock)

		const app = shallow(<App />)
		return app.instance().loadData().then(res=> {
			expect(app.state().notes).toBe(data_db_mock)
		})
	})

	it('should be able to save notes to database and load same data (real firebase API call)', () => {
	 	window.alert = jest.fn()
		const app = shallow(<App />)

		// store new data
		const notes = [{color: 'blue', text: 'hello blue'}, {color: 'grey', text: 'hi grey'}]
		app.setState({notes})
		return app.instance().saveData().then(res => {
			expect(res).toBe('success')
			expect(window.alert).toBeCalledWith('notes saved in database')
			
			// loading method that will check the amount of notes
			return app.instance().loadData().then(res => {
				expect(res).toBe('success')
				expect(app.state().notes).toEqual(notes)
			})
		})
	})

	it('should be able to change state color', () => {
		const app = mount(<App />)
		app.find('#select-color-picker').simulate('change', {target: {value: 'green'}}) // click on green radio button
		expect(app.state().selectedColor).toBe('green')
	})

	it('should add note with default color blue', () => {
		const app = mount(<App />)
		app.instance().handleClickAddNote()
		app.update()
		const note = app.find('.note').get(0)
		expect(note.props).toHaveProperty('style', {backgroundColor: 'lightblue'})
		app.unmount()
	})

	it('should render note with correct color', () => {
		const app = mount(<App />)
		app.setState({notes: [{color: 'grey', text: 'hello'}]}) // add one blue note
		const note = app.find('.note').get(0)
		expect(note.props).toHaveProperty('style', {backgroundColor: 'lightgrey'})
		app.unmount()
	})


	it('should add note with correct color after user select a color', () => {
		const app = mount(<App />)
		app.find('#select-color-picker').simulate('change', {target: {value: 'green'}}) // click on green radio button
		app.instance().handleClickAddNote()
		app.update()
		const note = app.find('.note').get(0)
		expect(note.props).toHaveProperty('style', {backgroundColor: 'lightgreen'})
		app.unmount()
	})

	it('should add note with the same text', () => {
		const app = mount(<App />)
		app.find('#input-new-note').simulate('change', {target: {value: 'take this note'}})
		app.instance().handleClickAddNote()
		const notes = [{color: 'blue', text: 'take this note'}] // assert text value on state, blue is default color
		expect(app.state().notes).toEqual(notes)
		
		app.update()
		expect(app.find('.note-text').text()).toBe('take this note') // assert text value on node
		app.unmount()
	})	
})
