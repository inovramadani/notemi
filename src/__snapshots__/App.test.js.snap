// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<div
  className="App"
>
  <h1
    className="title center"
  >
    Notemi
  </h1>
  <span
    className="btn"
    id="btn-save"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#2196f3",
      }
    }
  >
    Save
  </span>
  <div
    className="input-note"
  >
    <input
      id="input-new-note"
      onChange={[Function]}
      value=""
    />
    <span
      className="btn"
      id="btn-add"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#2196f3",
        }
      }
    >
      Add note
    </span>
  </div>
  <div
    className="color-picker"
  >
    <form
      id="select-color-picker"
      onChange={[Function]}
    >
      <span>
        <input
          checked={true}
          className="color-picker-item"
          name="color-picker"
          onChange={[Function]}
          type="radio"
          value="blue"
        />
        <span>
          blue
        </span>
      </span>
      <span>
        <input
          checked={false}
          className="color-picker-item"
          name="color-picker"
          onChange={[Function]}
          type="radio"
          value="grey"
        />
        <span>
          grey
        </span>
      </span>
      <span>
        <input
          checked={false}
          className="color-picker-item"
          name="color-picker"
          onChange={[Function]}
          type="radio"
          value="green"
        />
        <span>
          green
        </span>
      </span>
    </form>
  </div>
  <div
    className="notes"
  />
</div>
`;
